doc/error_handling.md

```diff
@@ -0,0 +1,285 @@
+## Logger
+The logger can be used similar to the streams in the C++ standard API.
+To select the log level, the corresponding logger has to be used, e.g. LogErr, LogWarn etc.
+
+```
+LogWarn() << "log message " << someValue << "log message continued";
+```
+
+## Error Handler
+The most general use case is the following
+```
+if(noError()) {
+    //handle the regular case
+} else {
+    auto callback = []() { //some error handling callback};
+    errorHandler(Error::kSOME_ERROR_CODE, callback, ErrorLevel::SEVERE);
+}
+```
+
+If no callback or error level are specified, the error level is assumed to be FATAL by default.
+```
+errorHandler(Error::kSOME_ERROR_CODE)
+```
+
+If no callback but an error level is desired, a nullptr has to be provided for the callback.
+```
+errorHandler(Error::kSOME_ERROR_CODE, nullptr, ErrorLevel::MODERATE);
+```
+We should consider changing the order of arguments in a future design (callback and additional arguments last). This can be done in a redesign were we provide additional information about the error location as well.
+
+## Expects and Ensures
+
+Assume func is part of an inner API and not supposed to be called with a nullptr. We may have used a reference here, this is just for illustration.
+In addition the value pointed to is assumed to be in the range (-1024, 1024). While we could check this everytime, this may not be necessary if we specify that the caller is responsible to ensure that these conditions hold.
+
+```
+int myAlgorithm(int* ptr) {
+    Expects(ptr!=nullptr);
+    //observe the order, we only dereference after the nullptr check
+    Expects(*ptr > -1024 && *ptr < 1024);
+
+    int intermediate = timesTwo(*ptr);
+    //this may not be necessary here to ensure that the next function call is valid,
+    //but it states our expectations clearly
+    Ensures(intermediate % 2 == 0);
+
+    int result = abs(intermediate);
+
+    Ensures(result % 2 == 0);
+    Ensures(result >= 0);
+    Ensures(result < 2048);
+
+    return result;
+}
+```
+Note that in the case of nullptr checks it may be an option to use references in arguments (or **not_null** if it is supposed to be stored since references are not copyable). It should be considered that not_null incurs a runtime cost, which may be undesirable.
+When Expects and Ensures are implemented to leave no trace in Release Mode, we do not incur a runtime cost using them.
+
+## cxx::expected
+This example checks the arguments and if they are valid proceeds to compute a result and returns it.
+Otherwise it creates an Error object from an erroCode and returns this.
+ 
+```
+std::expected<SomeType, Error>::func(Arg arg) {
+    int errorCode = checkArg(arg);
+    if(noError()) {
+        SomeType result = computeResult(arg);
+        // optionally do something with result
+        return result;
+    }
+    return Error(errorCode);
+}
+```
+The caller is responsible for handling (or propagating) the error.
+
+```
+auto result = func(arg);
+if(result.has_value()) {
```

Comment:
The current API for expected has `has_error()` rather than `has_value()`.

Reply:
I saw that too ... we need to change that.

Reply:
I leave it as is now if the API is going to change. In particular, this call and corresponding tests could just be added in this PR.